File: api/test/api/cnpj_test.exs
defmodule Api.CnpjTest do
  use ExUnit.Case, async: true

  alias Api.Cnpj

  # Vetores de ouro conferidos à mão pelo módulo 11 (ASCII − 48):
  #   12ABC34501DE  → DV 3 e 5   (exemplo canônico do Serpro, alfanumérico)
  #   112223330001  → DV 8 e 1   (11.222.333/0001-81, numérico clássico — retrocompatível)
  @alfanumerico "12ABC34501DE35"
  @numerico "11222333000181"

  describe "normalize/1" do
    test "mantém só [0-9A-Z] e sobe para maiúsculas" do
      assert Cnpj.normalize("12.abc.345/01de-35") == @alfanumerico
      assert Cnpj.normalize(" 11.222.333/0001-81 ") == @numerico
    end

    test "nil ou sem caractere válido viram nil" do
      assert Cnpj.normalize(nil) == nil
      assert Cnpj.normalize("") == nil
      assert Cnpj.normalize("  ./-  ") == nil
    end
  end

  describe "valid?/1" do
    test "aceita CNPJ alfanumérico válido, mascarado ou não" do
      assert Cnpj.valid?(@alfanumerico)
      assert Cnpj.valid?("12.ABC.345/01DE-35")
      assert Cnpj.valid?("12.abc.345/01de-35")
    end

    test "aceita CNPJ numérico clássico (mesma rotina)" do
      assert Cnpj.valid?(@numerico)
      assert Cnpj.valid?("11.222.333/0001-81")
    end

    test "rejeita dígito verificador errado" do
      refute Cnpj.valid?("12ABC34501DE34")
      refute Cnpj.valid?("11222333000182")
    end

    test "rejeita DV alfabético (as duas últimas posições são numéricas)" do
      refute Cnpj.valid?("12ABC34501DE3A")
    end

    test "rejeita tamanho diferente de 14 e entradas degeneradas" do
      refute Cnpj.valid?("")
      refute Cnpj.valid?("12ABC34501DE3")
      refute Cnpj.valid?("12ABC34501DE355")
      refute Cnpj.valid?("00000000000000")
      refute Cnpj.valid?(nil)
      refute Cnpj.valid?(123)
    end
  end
end

File: web/src/lib/cnpj.test.ts
import { describe, it, expect } from 'vitest';
import { normalizeCnpj, maskCnpj, isValidCnpj } from './cnpj';

// Vetores de ouro (conferidos pelo módulo 11 com ASCII − 48):
//   12ABC34501DE35 → alfanumérico válido (exemplo canônico do Serpro)
//   11222333000181 → numérico clássico válido (retrocompatível)

describe('normalizeCnpj', () => {
	it('mantém só [0-9A-Z] e sobe para maiúsculas', () => {
		expect(normalizeCnpj('12.abc.345/01de-35')).toBe('12ABC34501DE35');
		expect(normalizeCnpj(' 11.222.333/0001-81 ')).toBe('11222333000181');
	});
});

describe('maskCnpj', () => {
	it('aplica a máscara XX.XXX.XXX/XXXX-XX', () => {
		expect(maskCnpj('12ABC34501DE35')).toBe('12.ABC.345/01DE-35');
		expect(maskCnpj('11222333000181')).toBe('11.222.333/0001-81');
	});

	it('formata entradas parciais e trunca em 14', () => {
		expect(maskCnpj('12A')).toBe('12.A');
		expect(maskCnpj('12ABC345')).toBe('12.ABC.345');
		expect(maskCnpj('12ABC34501DE3599')).toBe('12.ABC.345/01DE-35');
		expect(maskCnpj('')).toBe('');
	});
});

describe('isValidCnpj', () => {
	it('aceita alfanumérico e numérico válidos, com ou sem máscara', () => {
		expect(isValidCnpj('12.ABC.345/01DE-35')).toBe(true);
		expect(isValidCnpj('12abc34501de35')).toBe(true);
		expect(isValidCnpj('11.222.333/0001-81')).toBe(true);
	});

	it('rejeita DV errado, tamanho errado e o CNPJ vazio', () => {
		expect(isValidCnpj('12ABC34501DE34')).toBe(false);
		expect(isValidCnpj('11222333000182')).toBe(false);
		expect(isValidCnpj('12ABC34501DE3')).toBe(false);
		expect(isValidCnpj('00000000000000')).toBe(false);
		expect(isValidCnpj('')).toBe(false);
	});
});
